### PR TITLE
feat(cache): Use the `CacheSpecificError` cache status

### PR DESCRIPTION
**Pull request by relaxolotl**
Welcome to the final part of my 4-part series to make gooder symbolicator cache statuses. This has a slightly larger number of requested reviewers than my other 3 PRs since it actually turns the switch on to use the new cache status that's being introduced. If you're interested, feel free to visit the other 3 PRs listed below to get a bit of an idea of what the diffs of those look like.

We are now at step 5 of this list:
1. Introduce the new cache status without changing the existing functionality, convert the new status to Malformed if needed (#509)
1. Update cache expiration strategy selection so different caches will expire entries triggered by cache-specific problems after a timeout that makes sense for the cache (#516) 
1. Extend both the new cache status and the Malformed cache status to accept an arbitrary string that describes the root cause of the issue (#510) 
1. Identify and split out code paths that should be using the new cache status without changing their return value (#511)
1. 👉 Begin actually using the new cache status by returning it in places identified in the previous step and writing + reading those to and from the cache (#512) 👈 you are here

### Changes
Unlike the other 3 PRs, this does make functional changes to symbolicator:
- When a download failure occurs, return the appropriate errors from the downloaders
- Handling the errors returned from the previous step and converting them into the correct cache status
- Stop coercing cache-specific error cache entries into malformed entries, and begin writing them as their own type
- ~~Updating cache-specific error cache entries to use the same expiration strategy as negative cache entries~~ This is no longer needed, already done in #516 and the behaviour is locked behind the previous bullet point
- Updating the cficache and symcache to handle cache-specific cache entries correctly 
- docs and notes

### Notable new behaviour
- Unlike Negative and Malformed, CacheSpecificError does not propagate from the data cache into the derived caches. It does propagate into the meta cache.

### CODE REVIEW

**Review by loewenheim — APPROVED** [review]

File: crates/symbolicator/src/services/objects/data_cache.rs [hunk 1]
```diff
@@ -348,10 +356,20 @@ mod tests {
...
             let result = objects_actor.find(find_object.clone()).await.unwrap();
-            assert_eq!(result.meta.unwrap().status, CacheStatus::Negative);
-            assert_eq!(server.accesses(), 1);
+            assert_eq!(
+                result.meta.unwrap().status,
+                CacheStatus::DownloadError(String::from(
+                    "failed to download: 500 Internal Server Error"
+                ))
+            );
+            assert_eq!(server.accesses(), 1 + 3); // 1 initial attempt + 3 retries
```
[the remark this answers is not in the document]
Comment (relaxolotl) [reply crates/symbolicator/src/services/objects/data_cache.rs:367]: done, thanks for calling this out!

File: crates/symbolicator/src/services/download/mod.rs [hunk 2]
```diff
@@ -57,6 +57,8 @@ pub enum DownloadError {
     #[allow(unused)]
```
[the remark this answers is not in the document]
Comment (relaxolotl) [reply crates/symbolicator/src/services/download/mod.rs]: done, thanks for catching this!

**Comment by relaxolotl** [comment]
rebased this and updated the changes to reflect the change in definition to the new enum.

as mentioned in the description,`CacheSpecificError` entries do not propagate from the data cache into all other caches.

other changes in the diff:
- i completely forgot about bitcode and updated its `compute` to cache download errors in the same way all of the other download caches do
- added more tests to make sure that all cache types use the correct cache entry expiration strategy for the new cache type
- updated tests + their docs
- changed the structure of the match in `from_derived_status` to see if it helps make it harder to miss any edge cases

**Review by loewenheim — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]

**Comment by Swatinem** [comment]
Same thing though as with the noperm changes: please test this manually to make sure sentry backend/ui can deal with these new candidate infos being returned.

**Comment by relaxolotl** [comment]
for posterity: checked my build locally and cache specific errors show up like so


Screenshot 2021-08-19 at 8 47 30 AM
